Sources/SwiftCompilerPlugin/Diagnostics.swift

```diff
@@ -0,0 +1,125 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftDiagnostics
+import SwiftSyntax
+
+/// Errors in macro handing.
+enum MacroExpansionError: String {
+  case macroTypeNotFound = "macro expanding type not found"
+  case freestandingMacroSyntaxIsNotMacro = "macro syntax couldn't be parsed"
+}
+
+extension MacroExpansionError: DiagnosticMessage {
+  var message: String {
+    self.rawValue
+  }
+  var diagnosticID: SwiftDiagnostics.MessageID {
+    .init(domain: "\(type(of: self))", id: "\(self)")
```

Comment:
Should we include the name of the module in the ID? We do that in the parser.

Reply:
Hm yeah, maybe `SwiftCompilerPlugin.MacroExpansionError` would make more sense. Is `domain` a user-visible value?

(fwiw, I just followed [`ParserDiagnosticMessages`](https://github.com/apple/swift-syntax/blob/main/Sources/SwiftParserDiagnostics/ParserDiagnosticMessages.swift#L81))

Reply:
Ah, I think I misread something, so do you think this should be `.init(domain: "MacroExpansionError", id: "\(self)")`, or anything else?

Reply:
Yes, exactly